### PR TITLE
[SFPA-61] Add ScannerView component

### PR DESCRIPTION
**Pull request by stoikokolev**

### CODE REVIEW

File: SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        
+        let request = VNRecognizeTextRequest { [weak self] request, error in
+            guard let observations = request.results as? [VNRecognizedTextObservation] else { return }
+            
+            observations.forEach { observation in
+                if let recognizedText = observation.topCandidates(1).first {
+                    self?.infoText.value = recognizedText.string
+                }
+            }
+        }
```
Comment (plamenzhl) [inline SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift:41]: Add empty line after closing bracket

**Review by plamenzhl — CHANGES_REQUESTED** [review]
Can we add an option to show/hide the segmented control that changes the Scanner view type (QRCode or text)?

File: SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    // MARK: - Private Functions
+    private func setupBinding() {
+        viewModel.infoText.sink { [weak self] text in
+            DispatchQueue.main.async {
+                self?.infoLabel.text = text
+            }
+        }
+    }
+    
+    private func createrSession() {
```
Comment (NNikolov1415) [inline SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift:91]: Perhaps this method was meant to be named `createSession()`?

File: SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+    // MARK: - Properties
+    let infoText = Observable<String?>(nil)
+    var title: String {
+        CategoryItem.scannerView.title
+    }
+    
+    // MARK: - Public Functions
+    func scan(_ image: UIImage, scanType: ScanType) {
+            switch scanType {
```
Comment (NNikolov1415) [inline SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift:21]: Could you check the indentation here? It appears odd in my browser.

**Review by kostadinzamanov — APPROVED** [review]
Looks good to me

**Review by NNikolov1415 — APPROVED** [review]

**Review by viktor-georgiev — CHANGES_REQUESTED** [review]

File: SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    
+    /// Starts scanning.
+    func startScan() {
+        viewModel.start()
+        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
+            guard let session = self?.captureSession,
+                  !session.isRunning else { return }
+            
+            session.startRunning()
+            DispatchQueue.main.async { [weak self] in
```
Comment (m-sapunova) [inline SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift:70]: [optional]

No need to weakify self here, as self is already weakly captured by the outer closure.

File: SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannnerViewModelDelegate.swift [hunk 5]
```diff
@@ -0,0 +1,17 @@
+//
+//  ScannnerViewModelDelegate.swift
+//  SFPortfolioApp
+//
+//  Created by Stoyko Kolev on 27.10.21.
+//
+
+import UIKit
+
+protocol ScannnerViewModelDelegate: AnyObject {
```
[the remark this answers is not in the document]
Comment (stoikokolev) [reply SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannnerViewModelDelegate.swift:10]: We follow the convention protocol to be declared in separate file in the directory where is used.

File: SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (stoikokolev) [reply SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewModel.swift]: Removed the scanning login from DemoViewModel.

File: SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift (hunk 4 above)
Comment (stoikokolev) [reply SFPortfolioApp/SFPortfolioApp/Components/ScannerView/ScannerView.swift]: Removed

File: SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewController.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        scannerView?.startScan()
+    }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        scannerView?.stopScan()
```
[the remark this answers is not in the document]
Comment (stoikokolev) [reply SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewController.swift:34]: Fixed

**Review by plamenzhl — APPROVED** [review]

**Review by m-sapunova — APPROVED** [review]

**Review by nProdanov — APPROVED** [review]

File: SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewController.swift [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    
+    private func setupScanner() {
+        scannerView = ScannerView.create(delegate: self)
+        scannerView?.expand(in: scannerContainer)
+        scannerView?.setup()
+    }
+    
+    private func setupBinding() {
+        viewModel.infoText.sink { [weak self] text in
+            DispatchQueue.main.async {
```
[the remark this answers is not in the document]
Comment (viktor-georgiev) [reply SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewController.swift:52]: Do you need to call it additionally on `DispatchQueue.main`?
I believe it is called there so you can just leave the code in the block.

[the remark this answers is not in the document]
Comment (stoikokolev) [reply SFPortfolioApp/SFPortfolioApp/Modules/DemoModule/ScannerDemoScene/ScannerDemoViewController.swift]: I had forgotten to remove this after making the _Observable_ changes.

**Review by viktor-georgiev — APPROVED** [review]

**Pull request merged**